File: python/single_double_fft.py
import os

import matplotlib.pyplot as plt
import numpy as np
import scipy.fftpack

cur_dir = os.path.dirname(os.path.abspath(__file__))
product_dir = os.path.join(os.path.dirname(cur_dir), "products")


def single_double_fft(fft_fn):

    size = 1024

    x_single = np.random.rand(size).astype(np.float32)
    x_double = x_single.astype(np.float64)
    x_diff = np.abs(x_single.astype(np.float64) - x_double)

    fig, axes = plt.subplots(2, 1, figsize=(16, 9))

    axes[0].plot(x_diff)
    axes[0].grid(True)
    axes[0].set_title("Difference between input")

    f_single = fft_fn(x_single)
    f_double = fft_fn(x_double)

    print((f"Data type after applying FFT to single"
           f" precision array: {f_single.dtype}"))
    print((f"Data type after applying FFT to double"
           f" precision array: {f_double.dtype}"))

    f_diff = abs(f_single.astype(np.complex128) - f_double)

    axes[1].plot(f_diff)
    axes[1].grid(True)
    axes[1].set_title("Difference between FFT output")

    func_module = f"{fft_fn.__module__}.{fft_fn.__name__}"

    fig.suptitle(f"Single vs Double precision FFT for {func_module}")
    fig.tight_layout(rect=[0, 0.03, 1, 0.95])
    fig.savefig(
        os.path.join(product_dir, f"single_double_fft.{func_module}.png"))


if __name__ == "__main__":
    single_double_fft(np.fft.fft)
    single_double_fft(scipy.fftpack.fft)
    plt.show()

File: matlab/single_double_fft.m
function single_double_fft ()
  % Determine whether Matlab's FFT implementation computes a true single precision
  % FFT, or if it returns a double precision array even when computing the FFT
  % of a single precision array.

  size = 1024;

  x_single = rand(size, 1, 'single');
  x_double = cast(x_single, 'double');
  x_diff = abs(cast(x_single, 'double') - x_double);

  ax1 = subplot(211);
  plot(ax1, x_diff);
  grid(ax1, 'on'); xlim([0, size]); title('Difference between input')

  f_single = fft(x_single);
  f_double = fft(x_double);

  fprintf('Data type after applying FFT to single precision array: %s\n', class(f_single));
  fprintf('Data type after applying FFT to double precision array: %s\n', class(f_double));

  f_diff = abs(cast(f_single, 'double') - f_double);

  ax2 = subplot(212);
  plot(ax2, f_diff);
  grid(ax2, 'on'); xlim([0, size]); title('Difference between FFT output')

  saveas(gcf, './../products/single_double_fft.png');

end
